src/modes/include/hardware/keystore.hpp

```diff
@@ -0,0 +1,344 @@
+ *               ....         16  0-15    data offset for large values
+ * ```
+ *
+ * Where the 16 least-significant bits are used by user to enumerate keys.
+ */
+enum class PrefixValues : uint32_t
+{
+  generalKeys = indexKeyMask | (0x0 << 29), ///< General keys (user)
+  managerKeys = indexKeyMask | (0x1 << 29), ///< Manager keys (per-mode)
+  privateKeys = indexKeyMask | (0x2 << 29), ///< Private keys (system-only)
+  invalidKeys = indexKeyMask | (0x3 << 29), ///< (reserved) for later use
+};
+
+/// \private Used for no group index (global to all groups and all modes)
+static constexpr uint32_t noGroupIndex = 15;
+
+/// \private Used for no mode index (local to group, global to its modes)
+static constexpr uint32_t noModeIndex = 31;
+
+/// \private Return indexed key given its parameters
+static constexpr uint32_t LMBD_INLINE
+indexKeyFor(PrefixValues prefix, uint32_t groupId, uint32_t modeId, uint32_t offset, uint32_t index)
+{
+  uint32_t rawPrefix = (uint32_t)prefix;
+
+  assert((rawPrefix << 3) == 0); // prefix on 2 bits
+  assert((groupId >> 4) == 0);   // groupId on 4 bits (0-14)
+  assert((modeId >> 5) == 0);    // modeId on 5 bits (0-30)
+  assert((offset >> 4) == 0);    // offset on 4 bits (0-15)
+  assert((index >> 16) == 0);    // index on 16 bits (0-65535)
+
+  index |= indexKeyMask;
+  return (rawPrefix << 29) | (groupId << 25) | (modeId << 20) | (offset << 16) | index;
+}
+
+/// \private Orchestrate interactions with the indexed key store
+template<typename _EnumTy, int _groupId, int _modeId, PrefixValues _prefix = PrefixValues::generalKeys> struct KeyStore
+{
+  using EnumTy = _EnumTy;
+  static constexpr uint32_t groupId = _groupId < 0 ? noGroupIndex : _groupId;
+  static constexpr uint32_t modeId = _modeId < 0 ? noModeIndex : _modeId;
+  static constexpr PrefixValues prefix = _prefix;
+
+  // it's magic!
+  static constexpr uint16_t storeIdKey = 0xafff;
+
+  static_assert(sizeof(EnumTy) == sizeof(uint16_t), "uint16_t enum only");
+  static_assert((groupId >> 4) == 0, "Group index must fit on 4 bits!");
+  static_assert((modeId >> 5) == 0, "Mode index must fit on 5 bits!");
+
+  // \private maps to ContextTy::KeyProxy::setValue()
+  template<EnumTy key, typename T> static inline void LMBD_INLINE setValue(T value)
+  {
+    using valueTy = std::remove_cv_t<std::remove_reference_t<T>>;
+    constexpr uint16_t rawKey = (uint16_t)key;
+
+    static_assert(storeIdKey != rawKey, "error: key can not match storeIdKey (0xafff)");
+
+    // simple case: value is uint32_t, store it as-is
+    if constexpr (std::is_same_v<valueTy, uint32_t>)
+    {
+      constexpr auto idx = indexKeyFor(prefix, groupId, modeId, 0, rawKey);
+      fileSystem::set_value(idx, value);
+
+      // small case: value is smaller than uint32_t, store it on a uint32_t
+    }
+    else if constexpr (sizeof(valueTy) <= sizeof(uint32_t))
+    {
+      uint32_t storage = 0;
+      memcpy(&storage, &value, sizeof(value));
```

Comment:
I'm 90% sure that in cpp we have better options than memcpy, but none comes to mind right now

Reply:
In a pure C++ fashion, doing `std::reinterpret_cast` is not the right way of doing object representation change, contrary to a common belief among C++ developers (I wonder where such misinformation comes from)

The modern way of doing that is [`std::bit_cast`](https://en.cppreference.com/w/cpp/numeric/bit_cast) from C++20 which is explicitly defined as:
```c++
template<class To, class From>
std::enable_if_t<
    sizeof(To) == sizeof(From) &&
    std::is_trivially_copyable_v<From> &&
    std::is_trivially_copyable_v<To>,
    To>
// constexpr support needs compiler magic
bit_cast(const From& src) noexcept
{
    static_assert(std::is_trivially_constructible_v<To>,
        "This implementation additionally requires "
        "destination type to be trivially constructible");
 
    To dst;
    std::memcpy(&dst, &src, sizeof(To));
    return dst;
}
```

We are C++17 here, and yes I've yolo-ed the `std::is_trivially_copyable` and `std::is_trivially_constructible` criteria in the aforementioned code, but 90% of the time it should be safe for the end-user to use as-is.

What I can do, to reassure the audience, is to integrate the `std::bit_cast` implementation quoted above in our `modes/include/tools.hpp` toolbox, with a link to the [cppreference associated page](https://en.cppreference.com/w/cpp/numeric/bit_cast) to be more explicit.